File: ruby/labs/beginner/lab2-collections/solution.rb
# Lab 2: Collections & Iteration - Complete Solution

class Contact
  attr_accessor :name, :email, :phone, :tags

  def initialize(name, email, phone, tags = [])
    @name = name
    @email = email
    @phone = phone
    @tags = tags
  end

  def to_s
    tags_str = @tags.empty? ? "" : " [#{@tags.join(', ')}]"
    "#{@name} <#{@email}> | Phone: #{@phone}#{tags_str}"
  end

  def has_tag?(tag)
    @tags.include?(tag)
  end
end

class ContactManager
  attr_reader :contacts

  def initialize
    @contacts = {}
  end

  def add_contact(contact)
    @contacts[contact.email] = contact
    puts "✓ Added: #{contact.name}"
  end

  def find_by_email(email)
    @contacts[email]
  end

  def count
    @contacts.length
  end

  def each_contact
    @contacts.each_value { |contact| yield contact }
  end

  def find_by_name(search_term)
    @contacts.values.select { |contact| contact.name.downcase.include?(search_term.downcase) }
  end

  def find_by_tag(tag)
    @contacts.values.select { |contact| contact.has_tag?(tag) }
  end

  def all_tags
    @contacts.values.flat_map(&:tags).uniq.sort
  end

  def all_names
    @contacts.values.map(&:name)
  end

  def email_to_name_map
    @contacts.transform_values(&:name)
  end

  def group_by_initial
    @contacts.values.group_by { |contact| contact.name[0].upcase }
  end

  def to_hash_array
    @contacts.values.map do |contact|
      { name: contact.name, email: contact.email, phone: contact.phone, tags: contact.tags }
    end
  end

  def tag_counts
    @contacts.values.reduce(Hash.new(0)) do |counts, contact|
      contact.tags.each { |tag| counts[tag] += 1 }
      counts
    end
  end

  def statistics
    total = count
    tags = all_tags

    puts "\nContact Manager Statistics:"
    puts "  Total contacts: #{total}"
    puts "  Total tags: #{tags.length}"
    puts "  Tags: #{tags.join(', ')}"

    puts "\n  Tag Distribution:"
    tag_counts.each do |tag, count|
      percentage = (count.to_f / total * 100).round(1)
      puts "    #{tag}: #{count} (#{percentage}%)"
    end
  end

  def find_where
    @contacts.values.select { |contact| yield contact }
  end

  def reject_where
    @contacts.values.reject { |contact| yield contact }
  end

  def any?
    @contacts.values.any? { |contact| yield contact }
  end

  def all?
    @contacts.values.all? { |contact| yield contact }
  end

  def sorted_by(attribute)
    @contacts.values.sort_by { |contact| contact.send(attribute) }
  end

  def list_all
    if @contacts.empty?
      puts "No contacts in the system."
    else
      puts "\nAll Contacts (#{count}):"
      sorted_by(:name).each_with_index do |contact, index|
        puts "  #{index + 1}. #{contact}"
      end
    end
  end
end

# Main Program
puts "=" * 70
puts " " * 20 + "CONTACT MANAGER"
puts "=" * 70

manager = ContactManager.new

# Add contacts
puts "\nAdding contacts..."
manager.add_contact(Contact.new("Alice Johnson", "alice@example.com", "555-0001", [:friend, :work]))
manager.add_contact(Contact.new("Bob Smith", "bob@gmail.com", "555-0002", [:friend, :hobby]))
manager.add_contact(Contact.new("Charlie Brown", "charlie@work.com", "555-0003", [:work, :client]))
manager.add_contact(Contact.new("Diana Prince", "diana@example.com", "555-0004", [:friend]))
manager.add_contact(Contact.new("Eve Adams", "eve@startup.io", "555-0005", [:work, :startup, :friend]))

# List all contacts
manager.list_all

# Search by name
puts "\n" + "=" * 70
puts "SEARCH BY NAME"
puts "=" * 70
results = manager.find_by_name("alice")
puts "\nSearching for 'alice':"
results.each { |c| puts "  → #{c}" }

# Search by tag
puts "\n" + "=" * 70
puts "SEARCH BY TAG"
puts "=" * 70
puts "\nContacts tagged with :work:"
manager.find_by_tag(:work).each { |c| puts "  → #{c}" }

puts "\nContacts tagged with :friend:"
manager.find_by_tag(:friend).each { |c| puts "  → #{c}" }

# Group by initial
puts "\n" + "=" * 70
puts "GROUPED BY INITIAL"
puts "=" * 70
manager.group_by_initial.each do |initial, contacts|
  puts "\n#{initial}:"
  contacts.each { |c| puts "  • #{c}" }
end

# Custom searches with blocks
puts "\n" + "=" * 70
puts "CUSTOM SEARCHES"
puts "=" * 70

puts "\nContacts with multiple tags:"
manager.find_where { |c| c.tags.length > 1 }.each { |c| puts "  → #{c}" }

puts "\nContacts with Gmail addresses:"
manager.find_where { |c| c.email.include?('@gmail.com') }.each { |c| puts "  → #{c}" }

has_startup = manager.any? { |c| c.has_tag?(:startup) }
puts "\nAny startup contacts? #{has_startup}"

# Statistics
puts "\n" + "=" * 70
puts "STATISTICS"
puts "=" * 70
manager.statistics

puts "\n" + "=" * 70

File: dart/labs/beginner/lab2-collections/solution.dart
// Lab 2: Collections & Iteration - Complete Solution

/// Represents a contact with name, email, phone, and tags.
class Contact {
  String name;
  String email;
  String phone;
  Set<String> tags;

  Contact({
    required this.name,
    required this.email,
    required this.phone,
    Set<String>? tags,
  }) : tags = tags ?? {};

  @override
  String toString() {
    final tagStr = tags.isEmpty ? '' : ' [${tags.join(', ')}]';
    return '$name <$email> | Phone: $phone$tagStr';
  }

  bool hasTag(String tag) {
    return tags.contains(tag);
  }
}

/// Manages a collection of contacts with search and analytics features.
class ContactManager {
  final Map<String, Contact> _contacts = {};

  void addContact(Contact contact) {
    _contacts[contact.email] = contact;
    print('✓ Added: ${contact.name}');
  }

  Contact? findByEmail(String email) {
    return _contacts[email];
  }

  int get count => _contacts.length;

  List<Contact> get allContacts => _contacts.values.toList();

  /// Iterate over all contacts with a callback
  void forEachContact(void Function(Contact) action) {
    _contacts.values.forEach(action);
  }

  /// Find contacts by name (partial match, case-insensitive)
  List<Contact> findByName(String searchTerm) {
    final lower = searchTerm.toLowerCase();
    return _contacts.values
        .where((contact) => contact.name.toLowerCase().contains(lower))
        .toList();
  }

  /// Find contacts with specific tag
  List<Contact> findByTag(String tag) {
    return _contacts.values
        .where((contact) => contact.hasTag(tag))
        .toList();
  }

  /// Get all unique tags across contacts
  Set<String> get allTags {
    return _contacts.values
        .expand((contact) => contact.tags)  // Flatten tags from all contacts
        .toSet();
  }

  /// Get list of all names
  List<String> get allNames {
    return _contacts.values.map((contact) => contact.name).toList();
  }

  /// Get Map of email => name
  Map<String, String> get emailToNameMap {
    return _contacts.map((email, contact) => MapEntry(email, contact.name));
  }

  /// Group contacts by first letter of name
  Map<String, List<Contact>> groupByInitial() {
    final grouped = <String, List<Contact>>{};
    
    for (final contact in _contacts.values) {
      final initial = contact.name[0].toUpperCase();
      grouped.putIfAbsent(initial, () => []).add(contact);
    }
    
    return grouped;
  }

  /// Get contacts as List of Maps (for export)
  List<Map<String, dynamic>> toMapList() {
    return _contacts.values.map((contact) => {
      'name': contact.name,
      'email': contact.email,
      'phone': contact.phone,
      'tags': contact.tags.toList(),
    }).toList();
  }

  /// Count contacts by tag using fold
  Map<String, int> tagCounts() {
    return _contacts.values.fold<Map<String, int>>(
      {},
      (counts, contact) {
        for (final tag in contact.tags) {
          counts[tag] = (counts[tag] ?? 0) + 1;
        }
        return counts;
      },
    );
  }

  /// Display statistics about contacts
  void statistics() {
    final total = count;
    final tags = allTags.toList()..sort();

    print('\nContact Manager Statistics:');
    print('  Total contacts: $total');
    print('  Total unique tags: ${tags.length}');
    print('  Tags: ${tags.join(', ')}');

    print('\n  Tag Distribution:');
    final counts = tagCounts();
    counts.forEach((tag, count) {
      final percentage = (count / total * 100).toStringAsFixed(1);
      print('    $tag: $count ($percentage%)');
    });
  }

  /// Find contacts matching a custom condition
  List<Contact> findWhere(bool Function(Contact) predicate) {
    return _contacts.values.where(predicate).toList();
  }

  /// Check if any contact matches condition
  bool any(bool Function(Contact) predicate) {
    return _contacts.values.any(predicate);
  }

  /// Check if all contacts match condition
  bool every(bool Function(Contact) predicate) {
    return _contacts.values.every(predicate);
  }

  /// Sort contacts by name (using cascade notation)
  List<Contact> sortedByName() {
    return allContacts..sort((a, b) => a.name.compareTo(b.name));
  }

  /// List all contacts sorted by name
  void listAll() {
    if (_contacts.isEmpty) {
      print('No contacts in the system.');
    } else {
      print('\nAll Contacts ($count):');
      final sorted = sortedByName();
      for (var i = 0; i < sorted.length; i++) {
        print('  ${i + 1}. ${sorted[i]}');
      }
    }
  }
}

// Main Program
void main() {
  print('=' * 70);
  print(' ' * 20 + 'CONTACT MANAGER');
  print('=' * 70);

  final manager = ContactManager();

  // Adding contacts (demonstrating Set literals for tags)
  print('\nAdding contacts...');
  manager.addContact(Contact(
    name: 'Alice Johnson',
    email: 'alice@example.com',
    phone: '555-0001',
    tags: {'friend', 'work'},
  ));
  manager.addContact(Contact(
    name: 'Bob Smith',
    email: 'bob@gmail.com',
    phone: '555-0002',
    tags: {'friend', 'hobby'},
  ));
  manager.addContact(Contact(
    name: 'Charlie Brown',
    email: 'charlie@work.com',
    phone: '555-0003',
    tags: {'work', 'client'},
  ));
  manager.addContact(Contact(
    name: 'Diana Prince',
    email: 'diana@example.com',
    phone: '555-0004',
    tags: {'friend'},
  ));
  manager.addContact(Contact(
    name: 'Eve Adams',
    email: 'eve@startup.io',
    phone: '555-0005',
    tags: {'work', 'startup', 'friend'},
  ));

  // List all contacts
  manager.listAll();

  // Search by name
  print('\n${'=' * 70}');
  print('SEARCH BY NAME');
  print('=' * 70);
  final results = manager.findByName('alice');
  print('\nSearching for \'alice\':');
  for (final contact in results) {
    print('  → $contact');
  }

  // Search by tag
  print('\n${'=' * 70}');
  print('SEARCH BY TAG');
  print('=' * 70);
  print('\nContacts tagged with \'work\':');
  for (final contact in manager.findByTag('work')) {
    print('  → $contact');
  }

  print('\nContacts tagged with \'friend\':');
  for (final contact in manager.findByTag('friend')) {
    print('  → $contact');
  }

  // Group by initial
  print('\n${'=' * 70}');
  print('GROUPED BY INITIAL');
  print('=' * 70);
  final grouped = manager.groupByInitial();
  grouped.forEach((initial, contacts) {
    print('\n$initial:');
    for (final contact in contacts) {
      print('  • $contact');
    }
  });

  // Custom searches with predicates
  print('\n${'=' * 70}');
  print('CUSTOM SEARCHES');
  print('=' * 70);

  print('\nContacts with multiple tags:');
  final multiTag = manager.findWhere((c) => c.tags.length > 1);
  for (final contact in multiTag) {
    print('  → $contact');
  }

  print('\nContacts with Gmail addresses:');
  final gmailContacts = manager.findWhere((c) => c.email.contains('@gmail.com'));
  for (final contact in gmailContacts) {
    print('  → $contact');
  }

  final hasStartup = manager.any((c) => c.hasTag('startup'));
  print('\nAny startup contacts? $hasStartup');

  final allHavePhone = manager.every((c) => c.phone.isNotEmpty);
  print('All contacts have phone numbers? $allHavePhone');

  // Demonstrate spread operator and collection literals
  print('\n${'=' * 70}');
  print('COLLECTION LITERALS & SPREAD OPERATOR');
  print('=' * 70);

  // Get all tags from first 3 contacts using spread
  final firstThreeEmails = ['alice@example.com', 'bob@gmail.com', 'charlie@work.com'];
  final combinedTags = <String>{
    for (final email in firstThreeEmails)
      if (manager.findByEmail(email) != null)
        ...manager.findByEmail(email)!.tags
  };
  print('\nCombined tags from first 3 contacts: ${combinedTags.join(', ')}');

  // Collection if - get only work contacts
  final workContacts = [
    for (final contact in manager.allContacts)
      if (contact.hasTag('work'))
        contact.name
  ];
  print('Work contacts: ${workContacts.join(', ')}');

  // Statistics
  print('\n${'=' * 70}');
  print('STATISTICS');
  print('=' * 70);
  manager.statistics();

  // Demonstrate cascade notation
  print('\n${'=' * 70}');
  print('CASCADE NOTATION EXAMPLE');
  print('=' * 70);
  
  // Create and configure a contact using cascades
  final newContact = Contact(
    name: 'Frank Miller',
    email: 'frank@example.com',
    phone: '555-0006',
  )
    ..tags.add('friend')
    ..tags.add('hobby')
    ..tags.add('gaming');
  
  print('\nCreated with cascade: $newContact');

  print('\n${'=' * 70}');
}
